### PR TITLE
Don't modify the DDG scope when analyzing the comprehension

### LINKED ISSUE #387: Language Server incorrectly marks variables as undefined in nested list comprehensions
_From @mawright on November 11, 2018 3:38_

## Environment data

- VS Code version: 1.28.2
- Extension version (available under the Extensions sidebar): 2018.10.1
- OS and version: Ubuntu 16.04
- Python version (& distribution if applicable, e.g. Anaconda): Anaconda 3.6.6
- Type of virtual environment used (N/A | venv | virtualenv | conda | ...): conda 
- Relevant/affected Python packages and their versions: N/A

## Actual behavior

screenshot from 2018-11-10 19-32-37

## Expected  behavior

Linter does not incorrectly mark variable as undeclared

## Steps to reproduce:
1. See screenshot

## Logs
Output for `Python` in the `Output` panel (`View`→`Output`, change the drop-down the upper-right of the `Output` panel to `Python`)

```
Starting Microsoft Python language server.
Microsoft Python Language Server version 0.1.60.0
Initializing for /home/matt/anaconda3/envs/learn/bin/python
```

Output from `Console` under the `Developer Tools` panel (toggle Developer Tools on under `Help`)

```
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:1409  INFO no standard startup: not a new window
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:4274 Overwriting grammar scope name to file mapping for scope source.python.
Old grammar file: file:///usr/share/code/resources/app/extensions/python/syntaxes/MagicPython.tmLanguage.json.
New grammar file: file:///home/matt/.vscode/extensions/magicstack.magicpython-1.1.0/grammars/MagicPython.tmLanguage
e.register @ /usr/share/code/resources/app/out/vs/workbench/workbench.main.js:4274
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] vscode-icons is active!
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Create file systemwatcher with pattern */python
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Create file systemwatcher with pattern */*/python
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Display locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Notify locators are locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Detected refreshing of Interpreters, Arg 1: {}, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Register Intepreter Watcher, Arg 1: undefined, Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Build the workspace interpreter watcher, Arg 1: undefined, Return Value: UNABLE TO DETERMINE VALUE
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Get language server folder name, , Return Value: "languageServer.0.1.60"
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Git extension API method 'getGitPath' is deprecated.
t.log @ /usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Hide locator refreshing progress, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: All locators have completed locating, , Return Value: 
/usr/share/code/resources/app/out/vs/workbench/workbench.main.js:282 [Extension Host] Python Extension: Checking whether locactors have completed locating, , Return Value: 
```

_Copied from original issue: Microsoft/vscode-python#3297_

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #387.

I'm not sure why the old code modified `ddg.Scope`. I'm not a DDG expert yet, but this is the only place where the scope gets explicitly changed by something other than the DDG class itself, which feels wrong (especially given that it doesn't put it back the way it was, instead changing it to some other scope entirely).

The code I'm removing was added in: https://github.com/Microsoft/PTVS/commit/861e1e4e3559d165615631733cfc3953292a52ff#diff-20d37dda723b616a828c08df50fc739b

If I'm missing something obvious here, let me know.

### CODE REVIEW

File: src/Analysis/Engine/Impl/AnalysisUnit.cs [hunk 1]
```diff
@@ -478,14 +478,9 @@ public ComprehensionAnalysisUnit(Comprehension node, PythonAst parent, Interpret
 
         internal override void AnalyzeWorker(DDG ddg, CancellationToken cancel) {
             var comp = (Comprehension)Ast;
-            var forComp = comp.Iterators[0] as ComprehensionFor;
 
-            if (forComp != null) {
-                // evaluate the 1st iterator in the outer scope
-                ddg.Scope = _outerUnit.InterpreterScope;
+            if (comp.Iterators[0] is ComprehensionFor forComp) {
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/AnalysisUnit.cs]: Is it guaranteed that `comp.Iterators` has at least one item?

Comment (jakebailey) [reply src/Analysis/Engine/Impl/AnalysisUnit.cs]: Presumably. The original code picked out `[0]` unconditionally too.

Comment (jakebailey) [reply src/Analysis/Engine/Impl/AnalysisUnit.cs]: Yeah, I think it should always have at least one thing, but we may need to think about what happens when there's _more_ than one thing. For example, "fun" list comprehension to flatten a 2D list is:

```python
l = [[1, 2], [3, 4]]
f = [i for l2 in l for i in l2]
```

At the point of this call, `comp.Iterators` contains two things. It's probably a better idea to iterate and assign each. I'll give it a test quick.

Comment (jakebailey) [reply src/Analysis/Engine/Impl/AnalysisUnit.cs]: Iterating or not doesn't seem to change anything, from what I can tell. With or without this scope fix, I can't hover over anything but `l` and get the right type.

**Comment by marvingreenberg** [comment]
Just linking the comment about a related issue, I made from the other case, in case it is any different.

https://github.com/Microsoft/vscode-python/issues/3297#issuecomment-479165089

**Comment by jakebailey** [comment]
That is #830, which will be fixed in #819. (This old PR is on the old codebase, which has been essentially rewritten and deleted.)

**Pull request merged**